flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java

```diff
@@ -156,9 +129,41 @@ public StreamInputProcessor(
 	}
 
 	public boolean processInput() throws Exception {
-		if (isFinished) {
-			return false;
+		initializeNumRecordsIn();
+
+		StreamElement recordOrMark = input.pollNext();
+		if (recordOrMark == null) {
+			input.isAvailable().get();
+			return input.isFinished();
 		}
+
+		processElement(recordOrMark);
+		return true;
+	}
+
+	private void processElement(StreamElement recordOrMark) throws Exception {
+		if (recordOrMark.isWatermark()) {
+			// handle watermark
+			statusWatermarkValve.inputWatermark(recordOrMark.asWatermark(), currentChannel);
```

Comment:
1. It seems missing some logic about assigning value to `currentChannel` ,  and  currently it is always - 1.
2.  I think that the channel index should be hidden in `NetworkInput`.

Reply:
True, there was a bug. However I'm not sure if we should squeeze more functionality to the `NetworkInput`. As it is now, it's a pretty small and nice class, that has a single responsibility - deserializing records, it would be nice to keep it as that. 

I was actually thinking about maybe latter adding another wrapper class, that would wrap `Input` and encapsulate watermark handling?

Reply:
In fact, I care about that the input channel index is exposed to input processors. As long as it's not exposed to the input processor (for example, encapsulating a class as you said), it's fine to me.